spec/lib/tasks/existing_firms_sign_up_task_spec.rb

```diff
@@ -1,107 +1,117 @@
+        allow(DateTime).to receive(:current).and_return(expected_invitation_sent_at)
 
-      before do
+        described_class.notify(stub_inviter, output)
+        expect(User.first.invitation_sent_at.to_datetime).to be_within(1.second).of(expected_invitation_sent_at)
+      end
+
+      it 'invites but does not send email' do
         ActionMailer::Base.deliveries = []
+        described_class.notify(stub_inviter, output)
+        expect(ActionMailer::Base.deliveries).to be_empty
       end
 
-      it 'mails principals that have no account' do
-        create :principal, email_address: 'ben@example.com', firm: create(:firm)
+      context 'creates a csv entry for principal' do
+        before do
+          described_class.notify(stub_inviter, output)
+        end
 
-        described_class.notify
+        it 'frn' do
+          CSV.parse(output.first) do |line_data|
+            expect(line_data[0].to_i).to eq(bill.firm.fca_number)
+          end
+        end
 
-        expect(emails_sent.size).to eq(2)
-        expect(emails_sent).to all(have_attributes(subject: 'Invitation instructions'))
+        it 'firm name' do
+          CSV.parse(output.first) do |line_data|
+            expect(line_data[1]).to eq(bill.firm.registered_name)
+          end
+        end
 
-        emails_sent.sort_by(&:to).tap do |sorted_emails_sent|
-          expect(sorted_emails_sent.first.to).to include('ben@example.com')
-          expect(sorted_emails_sent.last.to).to include(bill.email_address)
+        it 'principal name' do
+          CSV.parse(output.first) do |line_data|
+            expect(line_data[2]).to eq(bill.full_name)
+          end
+        end
+
+        it 'email' do
+          CSV.parse(output.first) do |line_data|
+            expect(line_data[3]).to eq(bill.email_address)
+          end
+        end
+
+        it 'invitation url' do
+          CSV.parse(output.first) do |line_data|
+            expected_output = stub_inviter.invitation_url(User.first)
+            expect(line_data[4]).to eq(expected_output)
+          end
         end
       end
 
-      it 'mails only principals for firms with no parent' do
+      it 'does not create output for trading names' do
         subsidiary = create(:firm, parent: bills_firm)
         create :principal, email_address: 'firm@example.com', firm: bills_firm
-        create :principal, email_address: 'subsidiary@example.com', firm: subsidiary
+        create :principal, email_address: 'tradingname@example.com', firm: subsidiary
 
-        described_class.notify
+        described_class.notify(stub_inviter, output)
 
-        expect(emails_sent.first.to).to include('firm@example.com')
+        expect(output.length).to eq(1)
+        CSV.parse(output.first) do |line_data|
+          expect(line_data[3]).to eq('firm@example.com')
+        end
       end
 
-      it 'does not mail principals thats firm is not registered' do
+      it 'output does not include details where the firm is not registered' do
```

Comment:
Fixed up a few more around this. Appreciating the focus on the language around testing :+1: